### DISCUSSION TITLE
unclear recursive unnnest when chaining regexp_split_to_table

### DISCUSSION THREAD

**Discussion by Jiehong** [duckdb/duckdb#11426] [General]
Hello,

Recently, I was looking at https://yusuf.fyi/posts/linux-text-manipulation, and I was wondering if I could do that with duckdb for fun.

So I created the following file named `sp-current`:

```
Album        Tea For The Tillerman (Remastered 2020)
AlbumArtist  Yusuf / Cat Stevens
Artist       Yusuf / Cat Stevens
Title        Wild World
```

And then I tried to extract things so as to produce `Wild World by Yusuf/Cat Stevens` as output.

However, this stopped pretty early with the following test:

```
duckdb -c "from read_text('sp-current') select regexp_split_to_table(content, '\n').regexp_split_to_array('  +')"
```

giving the following output:

```
┌──────────────────────────────────────────────────────────────────────┐
│ regexp_split_to_array(regexp_split_to_table("content", '\n'), '  +') │
│                              varchar[]                               │
├──────────────────────────────────────────────────────────────────────┤
│ [Album, Tea For The Tillerman (Remastered 2020)]                     │
│ [AlbumArtist, Yusuf / Cat Stevens]                                   │
│ [Artist, Yusuf / Cat Stevens]                                        │
│ [Title, Wild World]                                                  │
└──────────────────────────────────────────────────────────────────────┘
```

And I wanted to change the `regexp_split_to_array` to `regexp_split_to_table` like this:

```
duckdb -c "from read_text('sp-current') select regexp_split_to_table(content, '\n').regexp_split_to_table('  +')"
```

But this leads to the following error:

```
Binder Error: Nested UNNEST calls are not supported - use UNNEST(x, recursive := true) to unnest multiple levels
LINE 1: from read_text('sp-current') select regexp_spl...
```

This error tells me that `read_text` should be used with unnnest and recursive set to true, but it's not obvious where this could actually be used at all.

So I have 2 remarks/questions:

- this error message is difficult to act on, and actually is a consequence of `regexp_split_to_table` probably using `unnnest` itself internally. Could this error message be improved?
- how can I chain 2 `regexp_split_to_table` so that I get a table with the left side as a column and the right side as a column too?

I guess it would be nice if TSV files were supported as input (or "SSV", spaces separated values).

Lastly, maybe I'm missing an obvious way to solve this, and I'd be glad to be shown how you'd handle this :)

**Comment by cmdlineluser** [comment]
I think you're right about the error message.

I've not used the `to_table` / `to_array` functions, but perhaps this is useful:

```python
duckdb.sql("""
from (
   from read_text('sp-current')
   select unnest(
      [ line.str_split_regex('\s{2,}') for line in content.split('\n')[:-2] ]
   ) as split
)
select 
   split[1] as name,  
   split[2] as value
""")
```

```python
┌─────────────┬─────────────────────────────────────────┐
│    name     │                  value                  │
│   varchar   │                 varchar                 │
├─────────────┼─────────────────────────────────────────┤
│ Album       │ Tea For The Tillerman (Remastered 2020) │
│ AlbumArtist │ Yusuf / Cat Stevens                     │
│ Artist      │ Yusuf / Cat Stevens                     │
│ Title       │ Wild World                              │
└─────────────┴─────────────────────────────────────────┘
```


The list comprehension is "syntax sugar" for list `.apply()`

```python
content.split('\n').apply(line -> line.str_split_regex('\s{2,}'))
```

I'm not sure why they differ but [`regexp_extract()`](https://duckdb.org/docs/sql/functions/char.html#regexp_extractstring-pattern-name_list) allows you to return a struct 

```python
from read_text('sp-current')
select regexp_extract(content, '(\S+)\s+(.+)', ['name', 'value'])
```

```python
# ┌─────────────────────────────────────────────────────────────────────────────┐
# │ regexp_extract("content", '(\S+)\s+(.+)', main.list_value('name', 'value')) │
# │                   struct("name" varchar, "value" varchar)                   │
# ├─────────────────────────────────────────────────────────────────────────────┤
# │ {'name': Album, 'value': Tea For The Tillerman (Remastered 2020)}           │
# └─────────────────────────────────────────────────────────────────────────────┘
```


Whereas [`regexp_extract_all()`](https://duckdb.org/docs/sql/functions/char.html#regexp_extract_allstring-regex-group0) only allows you to return a single group at a time.

You could do 2 separate calls:

```python
from read_text('sp-current')
select 
   unnest(regexp_extract_all(content, '(\S+)\s+(.+)', 1)) as name,
   unnest(regexp_extract_all(content, '(\S+)\s+(.+)', 2)) as value
```

```python
# ┌─────────────┬─────────────────────────────────────────┐
# │    name     │                  value                  │
# │   varchar   │                 varchar                 │
# ├─────────────┼─────────────────────────────────────────┤
# │ Album       │ Tea For The Tillerman (Remastered 2020) │
# │ AlbumArtist │ Yusuf / Cat Stevens                     │
# │ Artist      │ Yusuf / Cat Stevens                     │
# │ Title       │ Wild World                              │
# └─────────────┴─────────────────────────────────────────┘
```

But perhaps there is a better way?

**Comment by Jiehong** [comment]
@cmdlineluser : thanks!

Should I open a defect for having a better error message?

**Comment by soerenwolfers** [comment]
https://duckdb.org/docs/data/csv/overview.html explains how to change the separator/delimiter and many more options for reading CSV/TSV

**Reply by Jiehong** [reply]
In this case, this is spaces delimited, so it doesn't work very well:

```
$ duckdb -c "from read_csv('sp-current', delim='\s+', header=false)"
Invalid Input Error: The delimiter option cannot exceed a size of 1 byte.
```

Multiple spaces lead to the same error message, so not very usable.
